### DISCUSSION TITLE
Error building Eclipse MOSAIC

### DISCUSSION THREAD

**Discussion by CiscoFrisco** [eclipse/mosaic#121] [Q&A]
Hello once again,

So I've been trying to connect a TraCI client to SUMO from an external python server. Since I didn't know to which port connect and I was getting some errors, I explored MOSAIC's repo to understand how you connect your TraciClient to SUMO. After noticing that you were finding a free port if the given port was 0, I changed the port to a random number in the runtime.json file just so that I could use that same port when connecting another client in the python server. However, I was still unable to connect to SUMO, so I checked SUMO's documentation.

From what I saw in SUMO's documentation ([here](https://sumo.dlr.de/docs/TraCI/Interfacing_TraCI_from_Python.html#controlling_the_same_simulation_from_multiple_clients)), to connect several clients to SUMO I need to provide the exact number of clients connecting in the sumo launch command with the argument --num-clients. 

After looking through the MOSAICs repo I noticed in the AbstractSumoAmbassador class the method [getProgramArguments](https://github.com/eclipse/mosaic/blob/8584928e1a2cea38641dfe14e82de8205abbecb3/fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/AbstractSumoAmbassador.java#L1381) containing a list of sumo arguments that I believe are used in the launch command. So I decided to simply add the argument --num-clients to that list. However, when building Eclipse MOSAIC I get the following error

```
[ERROR] Failed to execute goal on project mosaic-sumo: Could not resolve dependencies for project
org.eclipse.mosaic:mosaic-sumo:jar:21.1-SNAPSHOT:
Failed to collect dependencies at org.eclipse.sumo:libsumo:jar:1.9.2:
Failed to read artifact descriptor for org.eclipse.sumo:libsumo:jar:1.9.2: 
Could not transfer artifact org.eclipse.sumo:libsumo:pom:1.9.2 from/to repo.eclipse.org (https://repo.eclipse.org/content/repositories/sumo):
Transfer failed for https://repo.eclipse.org/content/repositories/sumo/org/eclipse/sumo/libsumo/1.9.2/libsumo-1.9.2.pom 503
Service Unavailable -> [Help 1]
```

This is a server error which makes me question: did you move these dependencies to another endpoint or is the service just down?

Best regards,
Francisco Filipe

**Comment by CiscoFrisco** [comment]
It works now. The service was just unavailable at the time. 

However, after adding the num-clients argument, as expected the SUMO server waits for two clients to connect before starting. Since I implemented an ambassador to communicate with the python server so that it can connect itself to the SUMO server, is there any interaction or method implemented/overwritten by MOSAIC so that I can connect to the SUMO server immediately after it starts?

Initially, I tried connecting as soon as the ambassador initialized, but it wasn't working as the SUMO server wasn't even started at that time. I tried a workaround since I didn't know any other solution by connecting to SUMO after receiving the first VehicleRegistration interaction. But that doesn't solve the problem as the server waits before even starting the simulation.

Any suggestion is welcome

**Reply by chengyuan0124** [reply]
Hello, I am currently have same work with you and would like to know how you setOrder for multiple clients? I know that python traci module implements .setOrder function to support multi-clients but I couldn't find this configuration in MOSAIC. Could you kindly share about this work?
